src/pyhf/optimize/mixins.py

```diff
@@ -0,0 +1,121 @@
+"""Helper Classes for use of automatic differentiation."""
+from .. import get_backend
+from .common import shim
+
+import logging
+
+log = logging.getLogger(__name__)
+
+
+class OptimizerMixin(object):
+    """Mixin Class to build optimizers."""
+
+    def __init__(self, **kwargs):
+        """
+        Create an optimizer.
+
+        Args:
+            maxiter (`int`): maximum number of iterations. Default is 100000.
+            verbose (`bool`): print verbose output during minimization. Default is off.
+        """
+        self.maxiter = kwargs.pop('maxiter', 100000)
+        self.verbose = kwargs.pop('verbose', False)
+        self._minimizer = None
+
+        if kwargs:
+            raise KeyError(
+                f"""Unexpected keyword argument(s): '{"', '".join(kwargs.keys())}'"""
+            )
+
+    def minimize(
+        self,
+        objective,
+        data,
+        pdf,
+        init_pars,
+        par_bounds,
+        fixed_vals=None,
+        return_fitted_val=False,
+        do_grad=False,
+        do_stitch=False,
+        method='SLSQP',
+        **kwargs,
+    ):
+        """
+        Find Function Parameters that minimize the Objective.
+
+        Args:
+            objective: objective function
+            data: observed data
+            pdf: model
+            init_pars: initial parameters
+            par_bounds: parameter boundaries
+            fixed_vals: fixed parameter values
+            return_fitted_val: return bestfit value of the objective
+            do_grad (`bool`): enable autodifferentiation mode. Default is off.
+            do_stitch (`bool`): enable splicing/stitching fixed parameter.
+            method: minimization routine
+            kwargs: other options to pass through to underlying minimizer
+
+        Returns:
+            bestfit parameters
+
+        """
+        tensorlib, _ = get_backend()
+        tv, fixed_values_tensor, objective_and_grad, init, bounds = shim(
+            objective,
+            data,
+            pdf,
+            init_pars,
+            par_bounds,
+            fixed_vals,
+            do_grad=do_grad,
+            do_stitch=do_stitch,
+        )
+
+        if do_grad:
+            wrapped_objective = lambda pars: objective_and_grad(pars)[0]
+            jac = lambda pars: objective_and_grad(pars)[1]
+        else:
+            wrapped_objective = objective_and_grad
+            jac = None
+
+        minimizer_kwargs = dict(
+            method=method, jac=jac, bounds=bounds, fixed_vals=[], **kwargs
+        )
+        if not do_stitch:
+            minimizer_kwargs.update(dict(fixed_vals=fixed_vals))
```

Comment:
I think it would be nice if `shim` returned `wrapped_objective `, `init`, `minimizer_kwargs` directly (and `postprocess` so we don't need access to `tv` etc directly

Reply:
done, via `stitch_pars` function.